gremlin-go/driver/connectionPool.go

```diff
@@ -46,97 +45,126 @@ type loadBalancingPool struct {
+			// Set the least used connection.
+			if leastUsed == nil || connection.activeResults() < leastUsed.activeResults() {
+				leastUsed = connection
+			}
 		}
-		pool.connections = pool.connections[:validIndex]
+	}
+	pool.connections = validConnections
 
-		// Create new connection if no valid connections were found in the pool or the least used connection exceeded
-		// the concurrent usage threshold while the pool still has capacity for a new connection
-		if leastUsed == nil ||
-			(leastUsed.activeResults() >= pool.newConnectionThreshold && len(pool.connections) < cap(pool.connections)) {
-			return pool.newConnection()
+	if leastUsed == nil {
+		// If no valid connection is found.
+		if len(pool.connections) >= cap(pool.connections) {
+			// Return error if pool is full and no valid connection was found (should not ever happen).
+			return nil, newError(err0105ConnectionPoolFullButNoneValid)
 		} else {
+			// Return new connection if no valid connection was found and pool has capacity.
+			return newConnection()
+		}
+	} else if leastUsed.activeResults() >= pool.newConnectionThreshold && len(pool.connections) < cap(pool.connections) {
+		// If the number of active results in our least used connection has reached the threshold
+		// AND our pool size has not reached the capacity, attempt to return a new connection.
+		newConnection, err := newConnection()
+		if err != nil {
+			// New connection creation failed; use existing grabbed least-used connection.
+			pool.logHandler.logf(Warning, poolNewConnectionError, err.Error())
 			return leastUsed, nil
 		}
+		return newConnection, nil
+	} else {
+		// If our pool is at capacity OR our least busy connection is less than the threshold, return the least used connection.
+		return leastUsed, nil
 	}
 }
 
-func (pool *loadBalancingPool) newConnection() (*connection, error) {
-	connection, err := createConnection(pool.url, pool.logHandler, pool.connSettings)
-	if err != nil {
-		return nil, err
-	}
-	pool.connections = append(pool.connections, connection)
-	return connection, nil
-}
+func newLoadBalancingPool(url string, logHandler *logHandler, connSettings *connectionSettings,
+	newConnectionThreshold int, maximumConcurrentConnections int, initialConcurrentConnections int) (connectionPool, error) {
+	var wg sync.WaitGroup
+	wg.Add(initialConcurrentConnections)
+	var appendLock sync.Mutex
 
-func newLoadBalancingPool(url string, logHandler *logHandler, connSettings *connectionSettings, newConnectionThreshold int,
-	maximumConcurrentConnections int) (connectionPool, error) {
 	pool := make([]*connection, 0, maximumConcurrentConnections)
-	initialConnection, err := createConnection(url, logHandler, connSettings)
-	if err != nil {
-		return nil, err
+	errorList := make([]error, 0, maximumConcurrentConnections)
+	for i := 0; i < initialConcurrentConnections; i++ {
+		go func() {
+			defer wg.Done()
+			connection, err := createConnection(url, logHandler, connSettings)
+
+			appendLock.Lock()
```

Comment:
may be better to use lock only when err != nil?

Reply:
No. since the errors slice append is also not concurrency safe.